### ISSUE TITLE
I have some problem about object query.

### ISSUE THREAD

**Issue by Zhong-Zi-Zeng** [facebookresearch/detr#596]
The last page of the original paper shows a simple code for DETR. The decoder's input is just a random value of size (100, 256). However, on your GitHub, I can't understand what you did about the object query. Why did you use the embedding layer's weight instead of a random value?

**Comment by AjibolaPy** [comment]
That's also my confusion. I guess instead of using random values, the embedding weights was used and resized. Maybe it's the same. But is it trainable?

**Comment by AjibolaPy** [comment]
That's also my confusion. I guess instead of using random values, the embedding weights was used and reshaped. Maybe it's the same. But is it trainable?. I'll appreciate an answer if you've gotten the answer

**Comment by Zhong-Zi-Zeng** [comment]
I have implemented DETR and found that embedding weights is more convenient than random values when building a model.

**Comment by AjibolaPy** [comment]
Okay, were they trainable weights?. Is it the same as doing 
<code> 
n=45
dim=1024
embedding=nn.Embedding(n, dim)
embedding(nn.ones_like(embedding.weight).
</code>.
Ones multiplying the weight has same effect same as zeros being added to the weight. Or am I missing something.

**Comment by AjibolaPy** [comment]
> I have implemented DETR and found that embedding weights is more convenient than random values when building a model.

After doing:
<code> embedding=nn.Embedding(45, 2)<br>weights=embedding.weight</code>
Did you keep requires_grad=True. This is just my confusion

**Comment by AjibolaPy** [comment]
> I have implemented DETR and found that embedding weights is more convenient than random values when building a model.

After doing:
embedding=nn.Embedding(45, 2)
<br>
weights=embedding.weight
Did you keep requires_grad=True. This is just my confusion

**Comment by AjibolaPy** [comment]
> I have implemented DETR and found that embedding weights is more convenient than random values when building a model.

After doing:
<code>
embedding=nn.Embedding(45, 2)
<br>
weights=embedding.weight
</code>
Did you keep requires_grad=True. This is just my confusion

**Comment by AjibolaPy** [comment]
> I have implemented DETR and found that embedding weights is more convenient than random values when building a model.

Thanks for helping.
Assuming:
<code>
embedding=nn.Embedding(45, 2)
<br>
 weights=embedding.weight.unsqueeze(1).repeat(1,3)
<code>
Did you keep the weights requires_grad=True. This is just my confusion

**Comment by Zhong-Zi-Zeng** [comment]
Yes, so that the gradient of embedding weight will have the same value.

**Comment by AjibolaPy** [comment]
Thanks for your answer.
 I should keep requires_grad set to 'True.' That means it will also be trained during backpropagation, and the value will change. This applies to the embedding weights, specifically the query embedding.

**Comment by AjibolaPy** [comment]
> Yes, so that the gradient of embedding weight will have the same value.

Thanks for your answer.
I should keep requires_grad set to 'True.' That means it will also be trained during backpropagation, and the value will change. This applies to the embedding weights, specifically the query embedding.

**Comment by AjibolaPy** [comment]
> Yes, so that the gradient of embedding weight will have the same value.

I think it's in (num_queries, batch_size, dim) and not (batch_size, num_queries, dim)

**Comment by Zhong-Zi-Zeng** [comment]
You are right!

**Comment by MLDeS** [comment]
> I have implemented DETR and found that embedding weights is more convenient than random values when building a model.

What do you mean with more convenient? Is it that the results are better? Because, as shown in the DETR colab notebook, if you use, `nn.Parameter(torch.rand(100, hidden_dim))` as the queries with 100 being the num_queries and updated all parameters of the model, it should still work because the `nn.parameter` has `requires_grad=True` by default, so it will still update the parameters, right?
